cmd/tailscale/main.go

```diff
@@ -370,16 +375,19 @@ func (a *App) runBackend(ctx context.Context) error {
 		case n := <-notifications:
 			exitWasOnline := state.ExitStatus == ExitOnline
 			if p := n.Prefs; p != nil && n.Prefs.Valid() {
-				state.Prefs = p.AsStruct()
+				// make a copy of prefs and use the readonly view
+				state.Prefs = p.AsStruct().View()
 				state.updateExitNodes()
-				a.setPrefs(state.Prefs)
+				a.prefs <- state.Prefs
 			}
-			first := state.Prefs == nil
+			first := !state.Prefs.Valid()
 			if first {
-				state.Prefs = ipn.NewPrefs()
-				state.Prefs.Hostname = a.hostname()
-				go b.backend.SetPrefs(state.Prefs)
-				a.setPrefs(state.Prefs)
+				newPrefs := ipn.NewPrefs()
+				newPrefs.Hostname = a.hostname()
+				go b.backend.SetPrefs(newPrefs)
+				state.Prefs = newPrefs.View()
+
+				go a.editPrefs(ctx, state.Prefs)
```

Comment:
Instead of starting this here lazily, it might be safer to just start it before we enter the for loop on line 359. Otherwise, if we get more than two notifications with a valid PrefsView, we'll end up deadlocking because we push them to a.prefs but it can only buffer one.

If you do that, then I think(?) you should be able to make a.prefs an unbuffered channel.